src/Collection.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+namespace Friendica;
+
+use Friendica\Database\Database;
+use Friendica\Database\DBA;
+use Psr\Log\LoggerInterface;
+
+/**
+ * The Collection classes inheriting from this abstract class are meant to represent a list of database record.
+ * The associated model class has to be provided in the child classes.
+ *
+ * Collections can be used with foreach().
+ */
+abstract class Collection implements \Iterator, \Countable, \ArrayAccess
```

Comment:
a general question: Why don't we use https://www.php.net/iteratoriterator and extend the missing parameters?

I got these idea from: https://www.schmengler-se.de/2017/04/php-7-typsichere-arrays-von-objekten/ which seems like almost the solution because of the same reasons :-) (badly it's in german only ...)

He's doing it like:
```php
class Users extends IteratorIterator
{
    public function __construct(User ...$users)
    {
        parent::__construct(new ArrayIterator($users));
    }
    public function current() : User
    {
        return parent::current();
    }
}
```

which would be, in our case:
```php
class Collection extends IteratorIterator
{
   /** @var int */
   protected $totalCount = 0;

    public function __construct(int $totalCount, Model ...$models)
    {
        parent::__construct(new ArrayIterator($models));
    }
    public function current() : Model
    {
        return parent::current();
    }
}
```
and we would get rid of the custom rewind/current methods which will be implemented already :-)

Reply:
Mainly because of simplicity. I've looked into the Iterator interfaces, but I didn't feel like they were bringing anything else to the table. We don't do any merge or sort on the collection yet.

> we would get rid of the custom rewind/current methods which will be implemented already

Is that so?

Reply:
It seems like, because it's already implemented by the `IteratorIterator` class, which we would extend. And maybe there are pitfalls in implementing the Iterator interfaces, which we maybe avoid with using a PHP "standard" class.

But it's up to you :-) It was just a suggestion

Reply:
Ooooh, I see what you mean, the `IteratorIterator` takes care of the `Collectiono->models` property. We would still need to implement `ArrayAccess` methods though.

Reply:
yep

Reply:
I finally went with `ArrayIterator` because it does implement `ArrayAccess`. If you read your article again, it suggests to use `IteratorIterator` when you **don't** need to access the collection like an array.